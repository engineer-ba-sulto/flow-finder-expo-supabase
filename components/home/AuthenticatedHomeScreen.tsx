import React from "react";
import { RefreshControl, ScrollView, Text, View } from "react-native";
import { Button } from "../../components/ui/Button";
import { SimpleGoalCompletion } from "../ui/SimpleGoalCompletion";
import { APP_DESCRIPTION } from "../../constants/app";

interface AuthenticatedHomeScreenProps {
  goalData: any;
  refreshing: boolean;
  onRefresh: () => void;
  fetchGoalCount: () => void;
  router: any;
  user: any;
}

const AuthenticatedHomeScreen: React.FC<AuthenticatedHomeScreenProps> = ({
  goalData,
  refreshing,
  onRefresh,
  fetchGoalCount,
  router,
  user,
}) => (
  <ScrollView
    className="flex-1 bg-gray-50"
    showsVerticalScrollIndicator={false}
    refreshControl={
      <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
    }
    accessibilityLabel="ホーム画面"
  >
    <View className="flex-1 px-4 pt-8 pb-6">
      {/* ウェルカムセクション */}
      <View className="bg-white rounded-xl p-6 mb-6 shadow-sm">
        <Text
          className="text-2xl font-bold text-center mb-2 text-gray-800"
          accessibilityRole="header"
          accessibilityLabel="Flow Finderへようこそ"
        >
          Flow Finderへようこそ
        </Text>
        <Text
          className="text-lg text-center mb-4 text-[#FFC400] font-medium"
          accessibilityLabel="おかえりなさい"
        >
          おかえりなさい
        </Text>
        {/* アプリ説明 */}
        <Text
          className="text-center text-gray-600 leading-6"
          accessibilityLabel={APP_DESCRIPTION}
        >
          {APP_DESCRIPTION}
        </Text>
      </View>
      {/* ゴール数表示セクション */}
      <View className="bg-white rounded-xl p-6 mb-6 shadow-sm">
        <Text
          className="text-lg font-semibold text-center mb-3 text-gray-800"
          accessibilityRole="header"
        >
          あなたの進捗
        </Text>
        {goalData.error ? (
          <View className="items-center">
            <Text
              className="text-red-500 text-center mb-3"
              accessibilityLabel={`ゴール数取得エラー: ${goalData.error}`}
            >
              {goalData.error}
            </Text>
            <Button
              variant="secondary"
              onPress={fetchGoalCount}
              accessibilityLabel="ゴール数を再取得"
              accessibilityHint="ゴール数の取得を再試行します"
            >
              再取得
            </Button>
          </View>
        ) : (
          <View className="bg-blue-50 rounded-lg p-4 items-center">
            <Text
              className="text-2xl font-bold text-blue-600 mb-1"
              accessibilityLabel={`登録ゴール数: ${goalData.count}件`}
            >
              {goalData.loading ? "..." : goalData.count}
            </Text>
            <Text
              className="text-blue-600 font-medium"
              accessibilityLabel="ゴール数の単位"
            >
              {goalData.loading ? "読み込み中" : "登録ゴール"}
            </Text>
          </View>
        )}
      </View>
      {/* 簡易ゴール完了機能セクション */}
      <View className="bg-white rounded-xl p-6 mb-6 shadow-sm">
        <SimpleGoalCompletion
          goalId="current-goal"
          isCompleted={false}
          onToggle={(goalId) => {
            console.log('ゴール完了切り替え:', goalId);
            // MVP2段目で実際の完了機能を実装予定
          }}
        />
      </View>

      {/* クイックアクションセクション */}
      <View className="bg-white rounded-xl p-6 shadow-sm">
        <Text
          className="text-lg font-semibold mb-4 text-gray-800"
          accessibilityRole="header"
        >
          クイックアクション
        </Text>
        <Button
          variant="primary"
          onPress={() => router.push("/(tabs)/goals")}
          accessibilityLabel="ゴール一覧に移動"
          accessibilityHint="登録されているゴールの一覧を確認できます"
          className="mb-3"
        >
          ゴールを確認
        </Button>
        <Button
          variant="secondary"
          onPress={() => router.push("/(tabs)/goals")}
          accessibilityLabel="新しいゴールを追加"
          accessibilityHint="新しいゴールの登録を開始します"
        >
          新しいゴールを追加
        </Button>
      </View>
    </View>
  </ScrollView>
);

export default AuthenticatedHomeScreen;
