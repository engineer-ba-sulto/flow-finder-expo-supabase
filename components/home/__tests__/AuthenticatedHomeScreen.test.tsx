import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react-native';
import AuthenticatedHomeScreen from '../AuthenticatedHomeScreen';

// モック設定
const mockRouter = {
  push: jest.fn(),
  replace: jest.fn(),
  back: jest.fn(),
};

jest.mock('../../../components/ui/Button', () => {
  const { Text, TouchableOpacity } = require('react-native');
  return {
    Button: ({ children, onPress, testID, className }: any) => (
      <TouchableOpacity onPress={onPress} testID={testID} className={className}>
        <Text>{children}</Text>
      </TouchableOpacity>
    ),
  };
});

jest.mock('../../../constants/app', () => ({
  APP_DESCRIPTION: 'あなたの成長を妨げる「見えない壁」を見つけ、壊すためのパーソナルコーチング アプリ',
}));

describe('AuthenticatedHomeScreen', () => {
  const defaultProps = {
    goalData: { count: 3, loading: false, error: null },
    refreshing: false,
    onRefresh: jest.fn(),
    fetchGoalCount: jest.fn(),
    router: mockRouter,
    user: { id: '1', email: 'test@example.com' },
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('基本表示', () => {
    it('ウェルカムメッセージが表示される', () => {
      render(<AuthenticatedHomeScreen {...defaultProps} />);
      
      expect(screen.getByText('Flow Finderへようこそ')).toBeTruthy();
      expect(screen.getByText('おかえりなさい')).toBeTruthy();
    });

    it('アプリ説明が表示される', () => {
      render(<AuthenticatedHomeScreen {...defaultProps} />);
      
      expect(screen.getByText('あなたの成長を妨げる「見えない壁」を見つけ、壊すためのパーソナルコーチング アプリ')).toBeTruthy();
    });

    it('進捗セクションが表示される', () => {
      render(<AuthenticatedHomeScreen {...defaultProps} />);
      
      expect(screen.getByText('あなたの進捗')).toBeTruthy();
    });

    it('クイックアクションセクションが表示される', () => {
      render(<AuthenticatedHomeScreen {...defaultProps} />);
      
      expect(screen.getByText('クイックアクション')).toBeTruthy();
      expect(screen.getByText('ゴールを確認')).toBeTruthy();
      expect(screen.getByText('新しいゴールを追加')).toBeTruthy();
    });
  });

  describe('ゴール数表示', () => {
    it('ゴール数が正常に表示される', () => {
      render(<AuthenticatedHomeScreen {...defaultProps} />);
      
      expect(screen.getByText('3')).toBeTruthy();
      expect(screen.getByText('登録ゴール')).toBeTruthy();
    });

    it('ローディング中は「...」と「読み込み中」が表示される', () => {
      const loadingProps = {
        ...defaultProps,
        goalData: { count: 0, loading: true, error: null },
      };
      
      render(<AuthenticatedHomeScreen {...loadingProps} />);
      
      expect(screen.getByText('...')).toBeTruthy();
      expect(screen.getByText('読み込み中')).toBeTruthy();
    });

    it('エラー時はエラーメッセージと再取得ボタンが表示される', () => {
      const errorProps = {
        ...defaultProps,
        goalData: { count: 0, loading: false, error: 'ゴール数の取得に失敗しました' },
      };
      
      render(<AuthenticatedHomeScreen {...errorProps} />);
      
      expect(screen.getByText('ゴール数の取得に失敗しました')).toBeTruthy();
      expect(screen.getByText('再取得')).toBeTruthy();
    });

    it('再取得ボタンをタップするとfetchGoalCountが呼ばれる', () => {
      const errorProps = {
        ...defaultProps,
        goalData: { count: 0, loading: false, error: 'ゴール数の取得に失敗しました' },
      };
      
      render(<AuthenticatedHomeScreen {...errorProps} />);
      
      const retryButton = screen.getByText('再取得');
      fireEvent.press(retryButton);
      
      expect(defaultProps.fetchGoalCount).toHaveBeenCalledTimes(1);
    });
  });

  describe('ナビゲーション', () => {
    it('「ゴールを確認」ボタンをタップするとゴール画面に遷移する', () => {
      render(<AuthenticatedHomeScreen {...defaultProps} />);
      
      const viewGoalsButton = screen.getByText('ゴールを確認');
      fireEvent.press(viewGoalsButton);
      
      expect(mockRouter.push).toHaveBeenCalledWith('/(tabs)/goals');
    });

    it('「新しいゴールを追加」ボタンをタップするとゴール画面に遷移する', () => {
      render(<AuthenticatedHomeScreen {...defaultProps} />);
      
      const addGoalButton = screen.getByText('新しいゴールを追加');
      fireEvent.press(addGoalButton);
      
      expect(mockRouter.push).toHaveBeenCalledWith('/(tabs)/goals');
    });
  });

  describe('プル更新', () => {
    it('refreshing=trueの時、RefreshControlが有効になる', () => {
      const refreshingProps = {
        ...defaultProps,
        refreshing: true,
      };
      
      const { toJSON } = render(<AuthenticatedHomeScreen {...refreshingProps} />);
      
      // RefreshControlのrefreshing プロパティが true になることを確認
      expect(toJSON()).toMatchSnapshot();
    });
  });

  describe('アクセシビリティ', () => {
    it('適切なアクセシビリティラベルが設定されている', () => {
      render(<AuthenticatedHomeScreen {...defaultProps} />);
      
      // アクセシビリティラベルが存在することを確認
      expect(screen.getByLabelText('ホーム画面')).toBeTruthy();
      expect(screen.getByLabelText('Flow Finderへようこそ')).toBeTruthy();
      expect(screen.getByLabelText('おかえりなさい')).toBeTruthy();
    });
  });
});