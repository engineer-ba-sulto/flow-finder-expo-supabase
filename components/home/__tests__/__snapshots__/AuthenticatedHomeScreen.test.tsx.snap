// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthenticatedHomeScreen プル更新 refreshing=trueの時、RefreshControlが有効になる 1`] = `
<RCTScrollView
  accessibilityLabel="ホーム画面"
  className="flex-1 bg-gray-50"
  refreshControl={
    <RefreshControlMock
      onRefresh={[MockFunction]}
      refreshing={true}
    />
  }
  showsVerticalScrollIndicator={false}
>
  <RCTRefreshControl />
  <View>
    <View
      className="flex-1 px-4 pt-8 pb-6"
    >
      <View
        className="bg-white rounded-xl p-6 mb-6 shadow-sm"
      >
        <Text
          accessibilityLabel="Flow Finderへようこそ"
          accessibilityRole="header"
          className="text-2xl font-bold text-center mb-2 text-gray-800"
        >
          Flow Finderへようこそ
        </Text>
        <Text
          accessibilityLabel="おかえりなさい"
          className="text-lg text-center mb-4 text-[#FFC400] font-medium"
        >
          おかえりなさい
        </Text>
        <Text
          accessibilityLabel="あなたの成長を妨げる「見えない壁」を見つけ、壊すためのパーソナルコーチング アプリ"
          className="text-center text-gray-600 leading-6"
        >
          あなたの成長を妨げる「見えない壁」を見つけ、壊すためのパーソナルコーチング アプリ
        </Text>
      </View>
      <View
        className="bg-white rounded-xl p-6 mb-6 shadow-sm"
      >
        <Text
          accessibilityRole="header"
          className="text-lg font-semibold text-center mb-3 text-gray-800"
        >
          あなたの進捗
        </Text>
        <View
          className="bg-blue-50 rounded-lg p-4 items-center"
        >
          <Text
            accessibilityLabel="登録ゴール数: 3件"
            className="text-2xl font-bold text-blue-600 mb-1"
          >
            3
          </Text>
          <Text
            accessibilityLabel="ゴール数の単位"
            className="text-blue-600 font-medium"
          >
            登録ゴール
          </Text>
        </View>
      </View>
      <View
        className="bg-white rounded-xl p-6 shadow-sm"
      >
        <Text
          accessibilityRole="header"
          className="text-lg font-semibold mb-4 text-gray-800"
        >
          クイックアクション
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <Text>
            ゴールを確認
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <Text>
            新しいゴールを追加
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
